### PR TITLE
SPP-10212-sml-google-analytics

### PR DESCRIPTION
**Pull request by jhmjam**
# Description

We want to use Google analytics to track site usage.

In order to use Google Analytics on our site we need to add the tags and update our cookies page to work with the new data gathering.

This PR covers adding the tags, updating the cookies page and updating the selenium tests.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Technical Enhancements

# Checklist:

If any of these are not completed, please explain why in the notes.

## **Definition of Done**
**Code and merges**

- [x] Code to be commented on where applicable 
- [x] Documentation updated where required 
- [x] Have considered non-functional requirements such as Security, Performance, Scalability, and Fault Tolerance
- [x] I have linted the code

**Testing**

- [x] All levels of acceptance test are passing (automated, integration, manual, accessibility, etc.)
- [x] I have run the behave command to check the selenium behaviour tests pass locally
- [x] Acceptance criteria met

**Other Checks**
- [x] I have performed a self-review of my own code
- [x] I have checked for spelling errors
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My changes don't break anything unexpected
- [x] I have checked and updated the security.txt file where required
- [x] Up to date with the main branch
- [x] For a change that needs to be deployed in a release I have a commit with the fix:, feat: or BREAKING CHANGE:
tags and will copy this commit message to the squash
- [ ] For a change that does not affect the deployed code I have added a commit with ci:, docs: or test:
depending upon which area of the solution was affected

Note when merging, squash commit must begin with one of the following tags:
"BREAKING CHANGE", "feat", "fix", "ci", "docs", "test"

Usage

BREAKING CHANGE: any change that breaks current functionality and may cause compatibility errors to users upgrading
    E.G BREAKING CHANGE: refactored method x to take variable y as input (changed from variable z)

feat: A new feature
    E.G feat: adding method x to future methods table

fix: A bug fix
    E.G fix: amending page for method x to show correct release version

ci: Changes to our CI configuration files and scripts (Concourse)

docs: Documentation only changes

test: Adding missing tests or correcting existing tests

### CODE REVIEW

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

File: sml_builder/static/js/cookies_function.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
+const DEFAULT_COOKIE_CONSENT = {
+    essential: true,
+    settings: false,
+    usage: false,
```
Comment (gibbardsteve) [inline sml_builder/static/js/cookies_function.js:4]: I think the default setting for the usage stuff be on. We won't get any tracking of users if the default is off.

File: sml_builder/static/js/cookies_function.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    campaigns: false,
+  };
+  
+const COOKIE_CATEGORIES = {
```
Comment (gibbardsteve) [inline sml_builder/static/js/cookies_function.js:8]: Where do these entries come from? Did you find some examples in the design system etc?   

I don't think we need all of these:
_ga_D60NB8YQQV: 'usage',
  _ga_F8PLZD05SG: 'usage',
  _ga_T2YY6YFBH2: 'usage',
  _ga_K9VRXZMXMJ: 'usage',
  _ga_WK76MC47: 'usage',
  
Are they the measurements IDs or the container IDs ?

File: sml_builder/templates/glossary.html [hunk 3]
```diff
@@ -82,4 +82,12 @@ <h3><a name="{{ term.hash }}"></a>{{ term.term }}</h3>
 
 {%- block scripts -%}
     <script type=text/javascript src="{{ url_for('static',filename='js/glossary.js') }}"></script>
+    <!-- Google Tag Manager -->
```
Comment (gibbardsteve) [inline sml_builder/templates/glossary.html:85]: This doesn't have the cookies code? I think we would always track the glossary as things stand.

File: sml_builder/templates/cookies.html [hunk 4]
```diff
@@ -8,161 +8,249 @@
...
-            "classes": "ons-u-mb-m ons-u-d-no ons-cookies-confirmation-message",
-            "type": "success",
-            "iconType": "check",
-            "iconSize": "xl"
-        }) -%}
-            <h2 id="cookie_change_alert">Your cookie settings have been saved</h2>
-            <p class="ons-u-mb-no">Some parts of the ONS may use additional cookies and will have their own cookie policy and banner.</p>
-            <a class="ons-js-return-link ons-u-mt-s ons-u-dib" href="#0">Return to previous page</a>
-        {%- endcall -%}
+<div class="ons-page__body">
```
Comment (gibbardsteve) [inline sml_builder/templates/cookies.html:11]: Formatting has changed quite a bit here. Have you run the tools that Joe has implemented on the pipeline? I think black etc is pretty strict about tabbing ets

File: features/steps/cookies.py [hunk 5]
```diff
@@ -84,26 +84,26 @@ def click_dropdown(context):
...
-def disable_cookie(context):
+@when('I change my settings to disable "{cookie_type}" cookies')
+def disable_cookie(context, cookie_type):
     WebDriverWait(driver, timeout=timeout).until(
-        lambda d: d.find_element(By.ID, "off-3-label")
+        lambda d: d.find_element(By.ID, f"off-{cookie_type}")
     ).click()
     WebDriverWait(driver, timeout=timeout).until(
-        lambda d: d.find_element(By.ID, "ons_cookie_save_changes")
+        lambda d: d.find_element(By.ID, "submit")
```
Comment (gibbardsteve) [inline features/steps/cookies.py:106]: Why is there a ID change here?

File: sml_builder/templates/cookies.html [hunk 6]
```diff
@@ -8,161 +8,249 @@
...
+                                    }
+                                ]
+                            }
+                        ]
                     })
                 }}
+            {% endcall %}
+            {{
+                onsButton({
+                    "id": 'submit',
```
Comment (gibbardsteve) [inline sml_builder/templates/cookies.html:235]: Can you check with @joepercival around these changes - I think there were some accessibility issues around descriptive text etc and you seem to have changed stuff here - I want to check this isn't undoing some of the work Joe has done.

**Review by gibbardsteve — COMMENTED** [review]
A few comments and questions against specific areas of the code.

File: sml_builder/static/js/cookies_function.js (hunk 1 above)
Comment (jhmjam) [reply sml_builder/static/js/cookies_function.js:4]: Done :)

File: sml_builder/static/js/cookies_function.js (hunk 2 above)
Comment (jhmjam) [reply sml_builder/static/js/cookies_function.js:8]: Hi yes these are the stream measurement id's and the container id. Its in relation to the cookie info found in the dev tools

Comment (jhmjam) [reply sml_builder/static/js/cookies_function.js:8]: I will see if we can leave them out :)

Comment (jhmjam) [reply sml_builder/static/js/cookies_function.js:8]: removed :)

File: sml_builder/templates/cookies.html (hunk 4 above)
Comment (jhmjam) [reply sml_builder/templates/cookies.html:11]: Hi ran it and no error was highlighted in relation to this but i have tabbed it anyway :)

File: features/steps/cookies.py (hunk 5 above)
Comment (jhmjam) [reply features/steps/cookies.py:106]: The new cookies.html has a different style of button. The 'type' of this button was labelled submit but did not have an ID. I added an Id and kept it consistent with the type but happy to change it now :)

Comment (jhmjam) [reply features/steps/cookies.py:106]: Done

File: sml_builder/templates/cookies.html (hunk 6 above)
Comment (jhmjam) [reply sml_builder/templates/cookies.html:235]: On it :)

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

File: sml_builder/templates/glossary.html (hunk 3 above)
Comment (jhmjam) [reply sml_builder/templates/glossary.html]: Hi Joe took a look and noted one thing but suspects its to do with the outdated version and thinks his changes will correct this :)

File: sml_builder/templates/cookies.html (hunk 6 above)
Comment (jhmjam) [reply sml_builder/templates/cookies.html]: Hi Joe took a look and noted one thing but suspects its to do with the outdated version and thinks his changes will correct this :)

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

File: sml_builder/templates/glossary.html (hunk 3 above)
Comment (jhmjam) [reply sml_builder/templates/glossary.html]: Hi Steve, so glossary does not seem to inherit the cookies code from core.html like the others even though it should. Just like the 'back to top' button does not work unless both scripts are added to the block script. This is tested and verified :)

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

File: sml_builder/static/js/cookies_function.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
+const DEFAULT_COOKIE_CONSENT = {
+    essential: true,
+    settings: false,
+    usage: true,
```
Comment (gibbardsteve) [inline sml_builder/static/js/cookies_function.js:4]: Sorry - having re-read the cookie policy in the design system I think it is correct that usage is false by default. The only cookie set by default should be essential. My misunderstanding of the context of the cookies policy.

File: sml_builder/static/js/cookies_function.js (hunk 2 above)
Comment (gibbardsteve) [reply sml_builder/static/js/cookies_function.js:8]: I tested on your branch and this seems to work ok, I was expecting we would need to add _ga_MeasurementID but it looks like it is working fine without.

File: sml_builder/templates/cookies.html [hunk 8]
```diff
@@ -36,47 +36,123 @@ <h2>Cookie settings</h2>
...
                     "title": "Learn more about these essential cookies",
-                    "titleTag": "h4",
+                    "titleTag": 'h4',
                     "button": {
                         "close": "Hide this"
                     }
-                }) -%}
-                    <h4>Your progress when filling in a survey or questionnaire</h4>
+                }) %}
+                    <h5 class="ons-u-fs-r--b">Your progress when filling in a survey or questionnaire</h5>
```
Comment (gibbardsteve) [inline sml_builder/templates/cookies.html:155]: We don't have any surveys or questionnaires so we should remove this.

File: sml_builder/templates/glossary.html (hunk 3 above)
Comment (gibbardsteve) [reply sml_builder/templates/glossary.html:85]: So I think we do need to fix this, it is not acceptable for us to ignore the cookie consent as it is part of GDPR etc. We need to find a way to add the cookie policy to the Glossary page, I don't think we can just say this page doesn't work like the others so we always track your usage.

**Review by gibbardsteve — COMMENTED** [review]
We need to find a fix for the glossary page.

File: sml_builder/static/js/cookies_function.js (hunk 7 above)
Comment (jhmjam) [reply sml_builder/static/js/cookies_function.js]: Reverted :)

File: sml_builder/templates/cookies.html (hunk 8 above)
Comment (jhmjam) [reply sml_builder/templates/cookies.html]: Removed :)

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://d1kauoyu9afulg.cloudfront.net/
cloudfront_id=EGHGUZJLLR5SO

**Review by gibbardsteve — APPROVED** [review]
Approved

**Pull request merged**